newrelic/hooks/mlmodel_langchain.py

```diff
@@ -255,6 +259,476 @@ def wrap_similarity_search(wrapped, instance, args, kwargs):
+    message_ids,
+):
+    expected_message_ids_len = len(input_message_list) + len(output_message_list)
+    actual_message_ids_len = len(message_ids)
+    if actual_message_ids_len < expected_message_ids_len:
+        message_ids.extend([str(uuid.uuid4()) for i in range(expected_message_ids_len - actual_message_ids_len)])
+        _logger.warning(
+            "The provided metadata['message_ids'] list was found to be %s when it needs to be at least %s. Internally generated UUIDs will be used in place of missing message ids."
+            % (actual_message_ids_len, expected_message_ids_len)
+        )
+
+    # Loop through all input messages received from the create request and emit a custom event for each one
+    for index, message in enumerate(input_message_list):
+        chat_completion_input_message_dict = {
+            "id": message_ids[index],
+            "appName": app_name,
+            "conversation_id": conversation_id,
+            "request_id": run_id,
+            "span_id": span_id,
+            "trace_id": trace_id,
+            "transaction_id": transaction.guid,
+            "content": message,
+            "completion_id": chat_completion_id,
+            "sequence": index,
+            "vendor": "langchain",
+            "ingest_source": "Python",
+            "virtual_llm": True,
+        }
+
+        transaction.record_custom_event("LlmChatCompletionMessage", chat_completion_input_message_dict)
+
+    if output_message_list:
+        # Loop through all output messages received from the LLM response and emit a custom event for each one
+        for index, message in enumerate(output_message_list):
+            # Add offset of input_message_length so we don't receive any duplicate index values that match the input message IDs
+            index += len(input_message_list)
+
+            chat_completion_output_message_dict = {
+                "id": message_ids[index],
+                "appName": app_name,
+                "conversation_id": conversation_id,
+                "request_id": run_id,
+                "span_id": span_id,
+                "trace_id": trace_id,
+                "transaction_id": transaction.guid,
+                "content": message,
+                "completion_id": chat_completion_id,
+                "sequence": index,
+                "vendor": "langchain",
+                "ingest_source": "Python",
+                "is_response": True,
+                "virtual_llm": True,
+            }
+
+            transaction.record_custom_event("LlmChatCompletionMessage", chat_completion_output_message_dict)
+
+
+def wrap_on_chain_start(wrapped, instance, args, kwargs):
+    run_manager = wrapped(*args, **kwargs)
+    transaction = current_transaction()
+    if not transaction:
+        return run_manager
+    # Only capture the first run_id.
+    if not hasattr(transaction, "_nr_run_manager_info"):
+        transaction._nr_run_manager_info = {
+            "run_id": run_manager.run_id,
```

Comment:
Should we safeguard here in case these attrs are not found on the `run_manager`?